### PR TITLE
MULTI-295: Add ability to create new canary deployment

### PR DESCRIPTION
**Pull request by EricGhildyal**

### CODE REVIEW

File: src/adapters/ingresses/apig.rs [hunk 1]
```diff
@@ -51,6 +106,15 @@ async fn read_file(artifact_path: PathBuf) -> Result<Vec<u8>> {
 #[async_trait]
 impl Ingress for AwsApiGateway {
     async fn deploy(&mut self) -> Result<()> {
-        todo!()
+        // First, we need to deploy the new version of the lambda
+        let lambda_version = self.upload_lambda("releases").await.unwrap();
```
Comment (RobbieMcKinstry) [inline src/adapters/ingresses/apig.rs:110]: In our next iteration on this code (not in this PR necessarily), it looks like we're going to need the release name as a configurable. Since it's required data for the ApiGateway struct to work, we should pass it in at construction time.

Same could be said for the `Monitor`'s configuration. I'll pick that up as part of MULTI-307 and MULTI-308.

Comment (RobbieMcKinstry) [inline src/adapters/ingresses/apig.rs:110]: OH! You mentioned just using `unwrap()` everywhere to solve you problems. Not so! I think you want to replace `unwrap` with `?` in most places. Certainly here!

Unwrap just crashes the program if there's an error. Instead, we want to return the error to the user. That's why this function takes a `Result` -- because it could succeeds with `Ok` or it could return an error.

The question mark operator functions as an early return. If your function returns `Result<T, E>`, and you suffix `?` after the value, the Rust will try to convert your value into `E`. It does this by calling `.into()` on your value. So as long as your value implements `Into<E>`, it will Just Work™️.

What complicates this: we're not using `std::result::Result`. We're using a fancy, colorful version called `miette::Result`, which is slightly different. Lots of errors in the std lib aren't directly compatible with `miette::Result`, but you can almost always upgrade them with `.into_diagnostic()`.

TL;DR: If you find yourself reaching for `unwrap`, first try `?`. If that doesn't work, try `.into_diagnostic()?`. Finally, if that doesn't work, try `.ok_or(Report::msg)`.

File: src/adapters/ingresses/apig.rs [hunk 2]
```diff
@@ -51,6 +106,15 @@ async fn read_file(artifact_path: PathBuf) -> Result<Vec<u8>> {
...
+
+        // Next, we need to create a new deployment, pointing at our
+        // new lambda version with canary settings
+        self.create_apig_deployment("Releases", "prod", "releases", &lambda_version, 10.0)
+            .await
+            .unwrap();
```
Comment (RobbieMcKinstry) [inline src/adapters/ingresses/apig.rs:116]: Same here

File: src/adapters/ingresses/apig.rs [hunk 3]
```diff
@@ -37,6 +37,61 @@ impl AwsApiGateway {
...
+
+        // Turn it into an uploadable zip file
+        let function_code = FunctionCode::builder().zip_file(code).build();
+
+        // Upload it to Lambda
+        let res = self
+            .lambda_client
+            .update_function_code()
+            .function_name(lambda_name)
+            .zip_file(function_code.zip_file().unwrap().clone())
```
Comment (RobbieMcKinstry) [inline src/adapters/ingresses/apig.rs:53]: Same here.

File: src/adapters/ingresses/apig.rs [hunk 4]
```diff
@@ -37,6 +37,61 @@ impl AwsApiGateway {
...
+            .send()
+            .await?;
+
+        Ok(res.version().unwrap().to_string())
```
Comment (RobbieMcKinstry) [inline src/adapters/ingresses/apig.rs:57]: This probably needs error handling too.

**Review by RobbieMcKinstry — APPROVED** [review]
Some changes I'd like for you to address before merge, mostly around error-handling of `unwrap()`, but overall LGTM.

File: src/adapters/ingresses/apig.rs [hunk 5]
```diff
@@ -51,6 +106,15 @@ async fn read_file(artifact_path: PathBuf) -> Result<Vec<u8>> {
 #[async_trait]
 impl Ingress for AwsApiGateway {
     async fn deploy(&mut self) -> Result<()> {
-        todo!()
+        // First, we need to deploy the new version of the lambda
+        let lambda_version = self.upload_lambda("releases").await.unwrap();
+
+        // Next, we need to create a new deployment, pointing at our
+        // new lambda version with canary settings
+        self.create_apig_deployment("Releases", "prod", "releases", &lambda_version, 10.0)
```
[the remark this answers is not in the document]
Comment (EricGhildyal) [reply src/adapters/ingresses/apig.rs:114]: Ahh, makes sense, thanks!
